pip/pip-439.md

```diff
@@ -0,0 +1,573 @@
+
+4. **Pending Acknowledge State**: Tracks message acknowledgments within transactions, preventing conflicts between competing transactions.
+
+## Transaction Lifecycle
+
+Transactions follow a defined lifecycle:
+
+1. **OPEN**: Client obtains a transaction ID from the Transaction Coordinator.
+2. **PRODUCING/ACKNOWLEDGING**: Client registers topic partitions/subscriptions with the TC, then produces/acknowledges messages within the transaction.
+3. **COMMITTING/ABORTING**: Client requests to end the transaction, TC begins two-phase commit.
+4. **COMMITTED/ABORTED**: After processing all partitions, TC finalizes the transaction state.
+5. **TIMED_OUT**: Transactions exceeding their timeout are automatically aborted.
+
+## Transaction Guarantees
+
+Pulsar transactions provide:
+- Atomic writes across multiple topics
+- Conditional acknowledgment to prevent duplicate processing by "zombie" instances
+- Visibility control ensuring consumers only see committed transaction messages
+- Support for exactly-once processing in consume-transform-produce patterns
+
+# Motivation
+
+Currently, Pulsar Functions cannot publish to multiple topics transactionally, which is a significant limitation for use cases requiring atomic multi-topic
+publishing. For instance, if a function processes an input message and needs to publish related updates to several output topics, there's no guarantee that all
+operations will succeed atomically.
+
+This limitation prevents building robust stream processing applications that require exactly-once semantics across multiple input and output topics. Without
+transaction support in Functions, developers must implement their own error handling and retry mechanisms, which can be complex and error-prone.
+
+Adding transaction support to Pulsar Functions would finally ensure message processing atomicity.
+
+# Goals
+
+## In Scope
+
+1. Enable automatic transaction support for Pulsar Functions through configuration
+2. Allow Functions to publish messages to multiple topics within a single transaction
+3. Support transactional acknowledgment of input messages
+4. Ensure transactions are committed only if message processing completes successfully
+5. Provide transaction timeout configuration for Functions
+
+## Out of Scope
+
+1. Exposing explicit transaction management APIs in the Functions interface
+2. Supporting multi-function transactions (transactions spanning multiple function invocations)
+3. Adding transaction support to Pulsar IO connectors
+4. Changes to the Function interface itself
+
+# High Level Design
+
+The proposed solution introduces automatic transaction wrapping for Pulsar Functions through configuration settings. When enabled, each function execution will be
+automatically wrapped in a transaction without requiring code changes to the function implementation.
+
+The general flow will be:
+1. Function is configured with `autoTransactionsEnabled: true`
```

Comment:
I'd suggest to replace the "auto transactions" concept with something that is more descriptive.
One possibility would be to call it "container managed transactions" or just "managed transactions". 

Since there's a plan to possibly support "explicit transactions" later, it would be better to use an enum instead of a boolean.


```java
    public enum TransactionMode {
        OFF,
        MANAGED
    }
```

There could be a TransactionConfig object to follow a similar pattern as there is for ProducerConfig:

```java
public class TransactionConfig {
  TransactionMode transactionMode;
  Long transactionTimeoutMs;
}
```

This would allow extending the configuration later if there becomes a need.